### PR TITLE
Distribute deployment in post commit tasks

### LINKED ISSUE #9964: Flaky `MultiPartitionDeploymentLifecycleTest.shouldTestLifecycle`
**Summary**

> Try to answer the following as best as possible

- How often does the test fail? More than once
- Does it block your work? No
- Do we suspect that it is a real failure? No

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
Error:  io.camunda.zeebe.engine.processing.deployment.MultiPartitionDeploymentLifecycleTest.shouldTestLifecycle  Time elapsed: 0.601 s  <<< FAILURE!
java.lang.AssertionError: 

Actual and expected have the same elements but not in the same order, at index 6 actual element was:
  (COMPLETED, EVENT)
whereas expected element was:
  (COMPLETE, COMMAND)

at io.camunda.zeebe.engine.processing.deployment.MultiPartitionDeploymentLifecycleTest.shouldTestLifecycle(MultiPartitionDeploymentLifecycleTest.java:59)
</pre>
</details>

**Hypotheses**

The ordering of events from multiple deployment distribution attempts is not guaranteed, the tests seems to be too strict.

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

This change makes it so that we do not distribute a deployment during the processing of the `CREATE` command, but as post commit task afters this command has been processed.

The reasoning behind this change is that the writing of the deployment distribution events/commands could happen in an unexpected order. This is because of the event buffering. Once we write an event this gets buffered. When we notify a different partition about the deployment it will write (and possibly process) the command immediately. This results in a situation where the different partition could write it's commands/events before we commands/events of the `CREATE` command have been written, making the ordering seem backwards.

E.g.:
1. We receive a `Deployment.CREATE` command
2. During processing we will write the `DeploymentDistribution.DISTRIBUTING` event. During processing we also send a message to the other partitions in order to distribute the deployment.
3. The `DeploymentDistribution.DISTRIBUTING` event is written to the buffer. Nothing is written to the log stream yet. The other partition receives the message and writes a `Deployment.DISTRIBUTE` command. At this point this partition is idle so it will immediately start processing this command.
4. Here we run into a race condition. If the second partition sends the response back to the first partition before the first partition has finished processing the `Deployment.CREATE` command it will write the `DeploymentDistribution.COMPLETE` before it writes the buffered events.

With this change the order of commands/events should always be the same. As a result this fixes the flaky test that is reference as a related issue. The flow will be as follows:

On partition 1
```
Command     Deployment.CREATE                      
Event       Process.CREATED                        
Event       Deployment.CREATED                     
Event       DeplyomentDistribution.DISTRIBUTING     (store distribution to partition N in state)
<Post commit tasks trigger at this point causing the other partitions to starts processing the deployment>
Command     DeploymentDistribution.COMPLETE         
Event       DeploymentDistribution.COMPLETED        
Event       Deployment.FULLY_DISTRIBUTEED           (when all partitions have triggered a complete command)

```

On other partitions
```
Command     Deployment.DISTRIBUTE       (causes the DeploymentDistribution.COMPLETE command to be written on partition 1)
Event       Deployment.DISTRIBUTED      
```

## Related issues

closes #9964 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [x] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/distribute/DeploymentDistributionBehavior.java [hunk 1]
```diff
@@ -54,7 +58,12 @@ public void distributeDeployment(final DeploymentRecord deploymentEvent, final l
           stateWriter.appendFollowUpEvent(
               key, DeploymentDistributionIntent.DISTRIBUTING, deploymentDistributionRecord);
 
-          distributeDeploymentToPartition(key, partitionId, copiedDeploymentBuffer);
+          sideEffectQueue.add(
+              () -> {
+                distributeDeploymentToPartition(key, partitionId, copiedDeploymentBuffer);
+                return true;
+              });
+          //          distributeDeploymentToPartition(key, partitionId, copiedDeploymentBuffer);
```
Comment (Zelldon) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/distribute/DeploymentDistributionBehavior.java:66]: :x: I guess this is a left-over :)

Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/distribute/DeploymentDistributionBehavior.java:66]: Haha oops 😄

**Comment by korthout** [comment]
>With this change the order of commands/events should always be the same. As a result this fixes the flaky test that is reference as a related issue.

@remcowesterhoud Please document this new order in the PR description.

**Comment by remcowesterhoud** [comment]
@korthout I have updated the description

**Review by korthout — COMMENTED** [review]
Thanks @remcowesterhoud 👏 Well researched!

❓ Would it be possible to write a test for this? 

🤔 I was thinking about a test where we assert the exact ordering using the recording log exporter

**Comment by remcowesterhoud** [comment]
@korthout You mean something like the [flaky test that caused me to fix this](https://github.com/camunda/zeebe/blob/main/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/MultiPartitionDeploymentLifecycleTest.java#L45-L102)? 😄

**Comment by korthout** [comment]
@remcowesterhoud Good point 😅

**Review by korthout — APPROVED** [review]
LGTM 🥇 

bors merge

**Pull request merged**